### PR TITLE
Support generation of last branch events with perf inject

### PR DESCRIPTION
**Pull request by sebpop**

### CODE REVIEW

**Comment by sebpop** [comment]
Mathieu, I just squashed the last patch in, and I removed the empty line.  Let me know if I need to change something else.
Thanks for your review!

File: tools/perf/util/cs-etm.c [hunk 1]
```diff
@@ -514,6 +538,65 @@ static int cs_etm__setup_queues(struct cs_etm_auxtrace *etm)
...
+
+static void cs_etm__update_last_branch_rb(struct cs_etm_queue *etmq,
+                                          struct cs_etm_packet *packet)
+{
+        struct branch_stack *bs = etmq->last_branch_rb;
+        struct branch_entry *be;
+
+        if (!etmq->last_branch_pos)
+                etmq->last_branch_pos = etmq->etm->synth_opts.last_branch_sz;
+
```
Comment (mathieupoirier) [inline tools/perf/util/cs-etm.c]: Add a comment that explains why ->last_branch_pos needs to be decremented by 1.

File: tools/perf/util/cs-etm.c [hunk 2]
```diff
@@ -514,6 +538,65 @@ static int cs_etm__setup_queues(struct cs_etm_auxtrace *etm)
...
+        etmq->last_branch_pos -= 1;
+
+        be              = &bs->entries[etmq->last_branch_pos];
+        be->from        = packet->start_addr;
+        be->to          = packet->end_addr;
+        /* No support for mispredict */
+        be->flags.mispred = 0;
+        be->flags.predicted = 1;
+
+        if (bs->nr < etmq->etm->synth_opts.last_branch_sz)
```
Comment (mathieupoirier) [inline tools/perf/util/cs-etm.c]: Are we always absolutely sure the ->nr will be behind by only one?

File: tools/perf/util/cs-etm.c [hunk 3]
```diff
@@ -809,8 +911,13 @@ static int cs_etm__sample(struct cs_etm_queue *etmq)
 
         err = cs_etm_decoder__get_packet(etmq->decoder,&packet);
         // if there is no sample, it returns err = -1, no real error
-
-        if (!err && packet.sample_type & CS_ETM_RANGE) {
+        if (err)
+                return err;
+        if (etmq->etm->synth_opts.last_branch &&
+            etmq->last_branch_rb->nr < etmq->etm->synth_opts.last_branch_sz) {
+                // FIXME: need to synth partial stack on last event.
```
Comment (mathieupoirier) [inline tools/perf/util/cs-etm.c]: I this work you intend to do in this patchset?  What is currently missing that prevents you from implementing?

File: tools/perf/util/cs-etm.c (hunk 2 above)
Comment (sebpop) [inline tools/perf/util/cs-etm.c]: bs->nr represents the number of branches already pushed to the bs (branch_stack).  It should be incremented until reaching the number of branches the user has specified in the command line "--itrace=il64", i.e., 64.  That number is made available in "etmq->etm->synth_opts.last_branch_sz".
Once bs->nr reaches that number it should remain set to that number: we then read that many entries from the branch stack and generate the last branch event.

**Comment by sebpop** [comment]
Mathieu,
I updated the patch and added more comments explaining how the circular buffers are managed.  I have removed the FIXME note as I implemented a way to drain the last branches from the circular buffer.
In reading the code again, I had to add two extra FIXME notes: one about trace sampling not working, and a question or assumption that I am making about the ETM traces, maybe you can help clearing that one out: does an ETM trace that we get from cs_etm__get_trace() only contain branches from a single cpu?

Thanks,
Sebastian

**Comment by mathieupoirier** [comment]
On 14 December 2016 at 15:21, sebpop <notifications@github.com> wrote:

> Mathieu,
> I updated the patch and added more comments explaining how the circular
> buffers are managed. I have removed the FIXME note as I implemented a way
> to drain the last branches from the circular buffer.
> In reading the code again, I had to add two extra FIXME notes: one about
> trace sampling not working, and a question or assumption that I am making
> about the ETM traces, maybe you can help clearing that one out: does an ETM
> trace that we get from cs_etm__get_trace() only contain branches from a
> single cpu?
>
Good morning,

When adding FIXME notes, please add a comment that explains what is
currently blocking the implementation.

As for "cs_etm__get_trace(), I haven't authored the original code and it's
been a long time since I looked at it but it's taking input from the
perf.data file itself so I would say that it can contain traces from any
CPU.

Mathieu

**Comment by mikel-armbb** [comment]
This pull request has been applied to the latest perf-opencsd code - and further update.
Now in https://github.com/Linaro/perf-opencsd repository, autoFDO branch
Closing pull request as it does not apply here any more

**Pull request closed without merging**